core/server/common/src/main/java/alluxio/master/journal/AsyncJournalWriter.java

```diff
@@ -99,49 +146,161 @@ public long appendEntry(JournalEntry entry) {
           mQueue.poll();
           writeCounter = mWriteCounter.incrementAndGet();
 
-          if (writeCounter >= targetCounter) {
-            if ((System.nanoTime() - startTime) >= mFlushBatchTimeNs) {
-              // This thread has been writing to the journal for enough time. Break out of the
-              // infinite for-loop.
-              break;
+          if ((System.nanoTime() - startTime) >= mFlushBatchTimeNs) {
+            // This thread has been writing to the journal for enough time. Break out of the
+            // infinite while-loop.
+            break;
+          }
+        }
+
+        if (needFlush) {
+          mJournalWriter.flush();
+          mFlushCounter.set(writeCounter);
+
+          /**
+           * Notify tickets that have been served to wake up.
+           */
+          List<FlushTicket> closedTickets = new ArrayList<>();
+          try (LockResource lr = new LockResource(mTicketLock)) {
+            for (FlushTicket ticket : mTicketList) {
+              if (ticket.getTargetCounter() <= writeCounter) {
+                ticket.setCompleted();
+                closedTickets.add(ticket);
+              }
             }
+            mTicketList.removeAll(closedTickets);
           }
         }
+      } catch (IOException | JournalClosedException exc) {
+        /**
+         * Release only tickets that have been flushed. Fail the rest.
+         */
+        List<FlushTicket> closedTickets = new ArrayList<>();
+        try (LockResource lr = new LockResource(mTicketLock)) {
+          for (FlushTicket ticket : mTicketList) {
+            closedTickets.add(ticket);
+            if (ticket.getTargetCounter() <= mFlushCounter.get()) {
+              ticket.setCompleted();
+            } else {
+              ticket.setError(exc);
+            }
+          }
+          mTicketList.removeAll(closedTickets);
+        }
+      }
+    }
+  }
+
+  /**
+   * Submits a ticket to flush thread and waits until ticket is served.
+   *
+   * If the specified counter is already flushed, this is essentially a no-op.
+   *
+   * @param targetCounter the counter to flush
+   */
+  @SuppressWarnings("Duplicates")
+  public void flush(final long targetCounter) throws IOException, JournalClosedException {
+    // Return if flushed.
+    if (targetCounter <= mFlushCounter.get()) {
+      return;
+    }
+
+    /**
+     * Submit the ticket for flush thread to process.
+     */
+    FlushTicket ticket = new FlushTicket(targetCounter);
+    try (LockResource lr = new LockResource(mTicketLock)) {
+      mTicketList.add(ticket);
+    }
+
+    try {
+      /**
+       * Notify the flush thread. It will have no effect if there is no ticket left to process,
+       * which means this ticket has been served as well.
+       */
+      synchronized (mFlushThread) {
+        mFlushThread.notify();
+      }
+
+      /**
+       * Wait on the ticket until completed.
+       */
+      ticket.waitCompleted();
```

Comment:
what happens if the flush thread flushes the entry BEFORE the ticket was added? Would the flush thread never notify this waiting thread?

Reply:
Do you mean flush thread never saw the ticket? If it saw it it'll mark it completed so waitCompleted() on it will be no-op.

Reply:
There was a window where flush was not causing flush thread to wake up. that should be fixed in latest as well.